File: DotNet/PopulationFitness/PopulationFitness/Models/Epochs.cs
using System;
using System.Collections.Generic;

namespace PopulationFitness.Models
{
    public class Epochs
    {
        public readonly List<Epoch> All;

        public Epochs()
        {
            All = new List<Epoch>();
        }

        /***
         * Adds an epoch with the specified start year.
         * Sets the end year of the preceding epoch to the year
         * before the start of this epoch (if there is a previous epoch)
         *
         * @param epoch
         */
        public void AddNextEpoch(Epoch epoch)
        {
            if (All.Count > 0)
            {
                Last.EndYear = epoch.StartYear - 1;
                epoch.PrevEnvironmentCapacity = Last.EnvironmentCapacity;
            }
            else
            {
                epoch.PrevEnvironmentCapacity = epoch.EnvironmentCapacity;
            }

            All.Add(epoch);
        }

        public void AddAll(List<Epoch> epochs)
        {
            foreach (Epoch e in epochs)
            {
                AddNextEpoch(e);
            }
        }

        /***
         * Set the end year of the readonly epoch
         *
         * @param last_year
         */
        public void SetFinalEpochYear(int last_year)
        {
            Last.EndYear = last_year;
        }

        /***
         *
         * @return the last epoch
         */
        public Epoch Last
        {
            get
            {
                return All[All.Count - 1];
            }
        }

        public Epoch First
        {
            get
            {
                return All[0];
            }
        }

        /**
         * Reduces the populations for all epochs by the same ratio
         *
         * P' = P/ratio
         *
         * @param ratio
         */
        public void ReducePopulation(int ratio)
        {
            foreach (Epoch epoch in All)
            {
                epoch.ReducePopulation(ratio);
            }
        }

        /**
         * Increases the populations for all epochs by the same ratio
         *
         * P' = P * ratio
         *
         * @param ratio
         */
        public void IncreasePopulation(int ratio)
        {
            foreach (Epoch epoch in All)
            {
                epoch.IncreasePopulation(ratio);
            }
        }

        public void PrintFitnessFactors()
        {
            foreach (Epoch epoch in All)
            {
                Console.Write("Epoch ");
                Console.Write(epoch.StartYear);
                Console.Write(" f=");
                Console.WriteLine(epoch.Fitness());
            }
        }
    }
}

File: Scala/PopulationFitness/src/main/scala/uk/edu/populationfitness/models/history/Epochs.scala
package uk.edu.populationfitness.models.history

import uk.edu.populationfitness.models.Config

import scala.collection.mutable.ListBuffer

class Epochs() {
  private val _epochs = new ListBuffer[Epoch]()

  def epochs: Seq[Epoch] = _epochs

  def years: Seq[EpochYear] = _epochs.flatMap(_.years)

  def config : Config = first.config

  def firstYear: Int = first.startYear

  def lastYear: Int = last.endYear

  def this(e: Seq[Epoch]){
    this
    _epochs ++= e
  }

  /** *
    * Adds an epoch with the specified start year.
    * Sets the end year of the preceding epoch to the year
    * before the start of this epoch (if there is a previous epoch)
    *
    * @param epoch
    */
  def add(epoch: Epoch): Unit = {
    if (!_epochs.isEmpty) {
      val previous = _epochs(_epochs.size - 1)
      previous.endYear = epoch.startYear - 1
      epoch.prevEnvironmentCapacity = previous.environmentCapacity
    }
    else{
      epoch.prevEnvironmentCapacity = epoch.environmentCapacity
    }
    _epochs += epoch
  }

  def addAll(epochs: Seq[Epoch]): Unit = {
    for (e <- epochs) {
      add(e)
    }
  }

  /** *
    * Set the end year of the final epoch
    *
    * @param last_year
    */
  def setFinalEpochYear(lastYear: Int): Unit = {
    last.endYear = lastYear
  }

  /** *
    *
    * @return the last epoch
    */
  def last: Epoch = _epochs(_epochs.size - 1)

  def first: Epoch = _epochs(0)

  /**
    * Reduces the populations for all epochs by the same ratio
    *
    * P' = P/ratio
    *
    * @param ratio
    */
  def reducePopulation(ratio: Int): Unit = {
    for (epoch <- epochs) {
      epoch.reducePopulation(ratio)
    }
  }

  /**
    * Increases the populations for all epochs by the same ratio
    *
    * P' = P * ratio
    *
    * @param ratio
    */
  def increasePopulation(ratio: Int): Unit = {
    for (epoch <- epochs) {
      epoch.increasePopulation(ratio)
    }
  }

  def printFitnessFactors(): Unit = {
    for (epoch <- epochs) {
      System.out.print("Epoch ")
      System.out.print(epoch.startYear)
      System.out.print(" f=")
      System.out.println(epoch.fitnessFactor)
    }
  }
}
